project-koku/masu
# Allow for current date to be adjusted in debug mode

**Overview**
During development and testing there are times when you would effectively want a time machine to go either forward or backwards to make sure reports are downloaded and processed for the 'current period'.  One way that closely aligns with real life is to adjust masu's concept of what the current day is.

By setting two environment variables `MASU_DEBUG=True` and `MASU_DATE_OVERRIDE=%Y-%m-%d %H:%M:%S` masu will initialize to the date and time set by `MASU_DATE_OVERRIDE`.

`/api/v1/status` enhancements were added so we will know whether or not masu is in debug mode and what it thinks the current date/time is.

***Important note for future development***
Any new masu logic that requires getting the current time or day i.e. `datetime.datetime.today() or datetime.datetime.now()` should use `DateAccessor().today()` instead.  This will ensure that masu behaves consistently.

This is a debug mode feature.  So I would encourage people to only use it when it's absolutely necessary.  Even though from the service point of view getting the current date will look identical to real life, it's still involves code only suitable for development (and some testing such as populating a database with historical data.  Once the database is populated, it's probably best to turn debug off)

**Testing**
1. Ensure MASU_DEBUG and MASU_DATE_OVERRIDE are unset.  Verify with /api/v1/status that current_datetime is today
2. Set MASU_DEBUG to true and MASU_DATE_OVERRIDE to remain as unset.  Verify with /api/v1/status that current_datetime is today
3. Set MASU_DEBUG to true and MASU_DATE_OVERRIDE to some other date.  Verify with /api/v1/status that current_datetime is adjusted to override date
4. Set MASU_DEBUG to false and MASU_DATE_OVERRIDE to some other date.  Verify with /api/v1/status that current_datetime is today's date
5. Check that /api/v1/status says DEBUG is true when env is set
6. Check that /api/v1/status says DEBUG is false when env is not set
7. Check that /api/v1/status says DEBUG is false when env is set to garbage

Results:
[masu_time_machine_ut.txt](https://github.com/project-koku/masu/files/2228514/masu_time_machine_ut.txt)
